### PR TITLE
Handle auto refreshing token

### PR DESCRIPTION
**Pull request by karolinakuzniewicz**

### CODE REVIEW

**Comment by patrys** [comment]
Task linked: [SALEOR-3987 Auto refresh login token](https://app.clickup.com/t/2549495/SALEOR-3987)

File: src/apollo/client.ts [hunk 1]
```diff
@@ -9,15 +9,74 @@ import {
...
+  email: string;
+  type: string;
+  user_id: string;
+  is_staff: boolean;
+};
+
 let client: ApolloClient<NormalizedCacheObject> | undefined;
 
-export const authLink = setContext((_, { headers }) => {
+const refreshToken = async (
```
Comment (jwm0) [inline src/apollo/client.ts:33]: This should be part of the SDK `auth` methods

File: package.json [hunk 2]
```diff
@@ -108,6 +108,6 @@
     "typescript": "^4.2.4"
   },
   "dependencies": {
-    "cross-fetch": "^3.1.4"
```
Comment (krzysztofwolski) [inline package.json:111]: Removing cross fetch is intentional? AFAIK its used by SSR

File: src/core/auth.ts [hunk 3]
```diff
@@ -149,6 +149,11 @@ export const auth = ({
       variables: { token },
     });
 
+    if (!result.data?.tokenVerify?.isValid) {
+      storage.setToken("");
+      console.log(storage.getToken());
```
Comment (krzysztofwolski) [inline src/core/auth.ts:154]: Is it left by mistake?

File: src/apollo/client.ts [hunk 4]
```diff
@@ -9,15 +9,74 @@ import {
...
+
+const autoRefreshFetch = async (
+  input: RequestInfo,
+  init: RequestInit
+): Promise<Response> => {
+  const initialRequest = fetch(input, init);
+  const token = storage.getToken();
+
+  if (JSON.parse(`${init.body}`).operationName === "refreshToken") {
+    return await initialRequest;
```
Comment (jwm0) [inline src/apollo/client.ts:61]:
```suggestion
    return initialRequest;
```

File: src/apollo/client.ts [hunk 5]
```diff
@@ -9,15 +9,74 @@ import {
...
+      await refreshToken((token: string) => {
+        init.headers = {
+          ...init.headers,
+          authorization: `JWT ${token}`,
+        };
+      });
+    }
+  }
+
+  return await initialRequest;
```
Comment (jwm0) [inline src/apollo/client.ts:77]:
```suggestion
  return initialRequest;
```

File: src/apollo/client.ts [hunk 6]
```diff
@@ -9,15 +9,74 @@ import {
...
+      storage.setToken(data.tokenRefresh.token);
+      callback(data?.tokenRefresh?.token);
+    }
+  } catch (error) {
+    storage.setToken("");
+    client!.resetStore();
+  }
+};
+
+const autoRefreshFetch = async (
```
Comment (jwm0) [inline src/apollo/client.ts:53]: This should also be exported out of SDK, could be used as a default fetch for other GraphQL clients

File: package.json (hunk 2 above)
Comment (karolinakuzniewicz) [reply package.json:111]: nope, it is also required for tests, removal was made by mistake.

File: src/core/auth.ts (hunk 3 above)
Comment (karolinakuzniewicz) [reply src/core/auth.ts]: removed

File: src/core/auth.ts [hunk 7]
```diff
@@ -149,6 +149,10 @@ export const auth = ({
       variables: { token },
     });
 
+    if (!result.data?.tokenVerify?.isValid) {
+      storage.setToken("");
```
Comment (jwm0) [inline src/core/auth.ts:153]:
```suggestion
      storage.setToken();
```

**Review by jwm0 — CHANGES_REQUESTED** [review]

File: src/core/auth.ts (hunk 7 above)
Comment (karolinakuzniewicz) [reply src/core/auth.ts:156]: `setToken` accepts only string and null

File: src/apollo/client.ts (hunk 6 above)
Comment (karolinakuzniewicz) [reply src/apollo/client.ts]: done

File: src/apollo/client.ts (hunk 5 above)
Comment (karolinakuzniewicz) [reply src/apollo/client.ts]: done

File: src/apollo/client.ts (hunk 4 above)
Comment (karolinakuzniewicz) [reply src/apollo/client.ts]: done

File: src/apollo/client.ts (hunk 1 above)
Comment (karolinakuzniewicz) [reply src/apollo/client.ts]: done

File: src/apollo/client.ts [hunk 8]
```diff
@@ -9,15 +9,74 @@ import {
...
 import { onError } from "@apollo/client/link/error";
 import { setContext } from "@apollo/client/link/context";
-import fetch from "cross-fetch";
+import jwtDecode from "jwt-decode";
 import { TypedTypePolicies } from "./apollo-helpers";
 import { REFRESH_TOKEN } from "./mutations";
 import { RefreshTokenMutation, RefreshTokenMutationVariables } from "./types";
 import { storage } from "../core/storage";
 
+type JWTToken = {
```
[the remark this answers is not in the document]
Comment (karolinakuzniewicz) [reply src/apollo/client.ts]: done

File: src/core/auth.ts (hunk 7 above)
Comment (krzysztofwolski) [reply src/core/auth.ts:156]: Then null is the choice

File: src/apollo/client.ts [hunk 9]
```diff
@@ -10,14 +10,49 @@ import {
...
+let client: ApolloClient<NormalizedCacheObject>;
+let refreshToken: AuthSDK["refreshToken"];
+
+export const autoRefreshFetch = async (
+  input: RequestInfo,
+  init: RequestInit
+): Promise<Response> => {
+  if (!client) {
+    throw new Error(
+      "Could not find saleor's apollo client in the context. Did you forget to call createSaleorClient()?"
```
Comment (jwm0) [inline src/apollo/client.ts:28]:
```suggestion
      "Could not find Saleor's client instance. Did you forget to call createSaleorClient()?"
```

**Review by jwm0 — APPROVED** [review]

**Comment by krzysztofwolski** [comment]
Jest tests are failing too

**Review by jwm0 — APPROVED** [review]

**Pull request merged**